### PR TITLE
http fix for infinite timeout disabling timeout infrastructure (#20817)

### PR DESCRIPTION
**Pull request by gosubpl**
- timeout infrastructure is no longer disabled when request-timeout is set to infinite

### CODE REVIEW

**Comment by gosubpl** [comment]
Refs #20817

**Comment by ktoso** [comment]
OK TO TEST

**Comment by gosubpl** [comment]
Fails locally too. How could I miss that? Looking into it.

**Comment by ktoso** [comment]
You're online quite early :)
Thanks in advance!

**Comment by gosubpl** [comment]
Needed to filter out `timeout-access` header in spec, as now it is attached to every request (so even when infinite timeout is specified)

**Comment by gosubpl** [comment]
Missed one spec in previous commit

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala [hunk 1]
```diff
@@ -307,13 +304,21 @@ private[http] object HttpServerBluePrint {
...
-      requestEnd.fast.map(_ ⇒ new TimeoutSetup(Deadline.now, schedule(initialTimeout, this), initialTimeout, this))
+    initialTimeout match {
+      case timeout: FiniteDuration ⇒ set {
+        requestEnd.fast.map(_ ⇒ new TimeoutSetup(Deadline.now, schedule(timeout, this), timeout, this))
+      }
+      case _ ⇒ set {
+        requestEnd.fast.map(_ ⇒ new TimeoutSetup(Deadline.now, new Cancellable {
+          override def isCancelled: Boolean = true
+          override def cancel(): Boolean = true
+        }, Duration.Inf, this))
```
Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: Could this `Cancellable` be a singleton object instead to save the allocation on each request?

**Comment by johanandren** [comment]
LGTM!

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala (hunk 1 above)
Comment (gosubpl) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: yeah, that's clumsy, sorry for that.
Will fix.

File: akka-docs/rst/scala/code/docs/http/scaladsl/server/directives/TimeoutDirectivesExamplesSpec.scala [hunk 2]
```diff
@@ -36,7 +34,7 @@ class TimeoutDirectivesExamplesSpec extends AkkaSpec(TimeoutDirectivesTestConfig
   import system.dispatcher
   implicit val materializer = ActorMaterializer()
 
-  def slowFuture(): Future[String] = Promise[String].future // move to Future.never in Scala 2.12
+  def slowFuture(): Future[String] = Promise[String].future // TODO: move to Future.never in Scala 2.12
```
Comment (ktoso) [inline akka-docs/rst/scala/code/docs/http/scaladsl/server/directives/TimeoutDirectivesExamplesSpec.scala]: thanks

File: akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala [hunk 3]
```diff
@@ -447,7 +447,7 @@ class ClientServerSpec extends WordSpec with Matchers with BeforeAndAfterAll wit
         val serverInSub = serverIn.expectSubscription()
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala]: Could you add a test that this PR directly makes pass @gosubpl ?
I.e. that we disabled it in config, but enabled in the routing layer, thus it should trigger properly

**Comment by ktoso** [comment]
I'd like one more test to cover this, as commented.
Otherwise LGTM

File: akka-docs/rst/scala/code/docs/http/scaladsl/server/directives/TimeoutDirectivesExamplesSpec.scala [hunk 4]
```diff
@@ -25,9 +25,7 @@ private[this] object TimeoutDirectivesTestConfig {
     akka.stdout-loglevel = ERROR
     windows-connection-abort-workaround-enabled = auto
     akka.log-dead-letters = OFF
-    akka.http.server.request-timeout = 1000s""")
-  // large timeout - 1000s (please note - setting to infinite will disable Timeout-Access header
-  // and withRequestTimeout will not work)
+    akka.http.server.request-timeout = infinite""")
```
Comment (gosubpl) [inline akka-docs/rst/scala/code/docs/http/scaladsl/server/directives/TimeoutDirectivesExamplesSpec.scala]: Before this PR, setting the `request-timeout` to `infinite` would result in some of the tests in `TimeoutDirectivesExamplesSpec` failing. In order to pass the timeout would have to be set to any finite value. Now the tests pass, independent of whether the `request-timeout` is finite (e.g. `1000s`) or `infinite`.

File: akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala (hunk 3 above)
Comment (gosubpl) [inline akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala]: @ktoso I have added a line comment above clarifying that the setting of `request-timeout` to `infinite` in line 28 of `TimeoutDirectivesExamplesSpec` before this PR would result in failed test. However, emboldened by your comment, I will add the opposite case (finite timeout, as was previously set) to pinpoint the fact, that now the setting is symmetric - works for both cases. The side effect of this change, which resulted in change in other specs, is that the `request-timeout` is never disabled in the routing layer but just set to infinite value. So we are changing the meaning of `request-timeout=infinite` from (special) _disable in routing layer_ to (plain) _set the timeout to infinite_.

**Comment by gosubpl** [comment]
Added test for finite timeout and a few (hopefully clarifying) comments.

File: akka-docs/rst/scala/code/docs/http/scaladsl/server/directives/TimeoutDirectivesExamplesSpec.scala (hunk 4 above)
Comment (ktoso) [inline akka-docs/rst/scala/code/docs/http/scaladsl/server/directives/TimeoutDirectivesExamplesSpec.scala]: excellent, that's what I had in mind :)

**Comment by ktoso** [comment]
Excellent PR, thanks @gosubpl !

**Comment by gosubpl** [comment]
Many thanks @johanandren @ktoso for assistance!
@ktoso, I would be grateful if you could take a look at my comment in https://github.com/akka/akka/issues/19361

**Pull request merged**